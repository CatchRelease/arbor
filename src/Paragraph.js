import styled, { css } from 'react-emotion';
import PropTypes from 'prop-types';

import { brandFont, colors, lineHeights, typography } from './theme';

const baseStyles = css`
  font-family: ${brandFont};
  line-height: ${lineHeights.large};
`;

const variantStyles = ({ color, variant }) => css`
  color: ${color ? colors[color] : typography.text[variant].color};
  font-size: ${typography.text[variant].fontSize};
  margin-bottom: ${typography.text[variant].fontSize};
`;

const withoutMarginStyles = ({ withoutMargin }) =>
  withoutMargin &&
  css`
    margin-bottom: 0;
  `;

const Paragraph = styled.p`
  ${baseStyles};
  ${variantStyles};
  ${withoutMarginStyles};
`;

Paragraph.propTypes = {
  color: PropTypes.oneOf(Object.keys(colors).concat('')),
  variant: PropTypes.oneOf(['ui', 'longForm', 'tiny']).isRequired,
  withoutMargin: PropTypes.bool
};

Paragraph.defaultProps = {
  color: '',
  withoutMargin: false
};

export default Paragraph;
