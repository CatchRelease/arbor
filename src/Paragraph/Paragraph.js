import { css } from '@emotion/core';
import styled from '@emotion/styled';
import PropTypes from 'prop-types';

import Text from '../Text';

const VARIANT_MAPPING = {
  ui: 'size4',
  longForm: 'size5',
  tiny: 'size2'
};

const variantStyles = ({ theme, variant }) => css`
  font-size: ${theme.fontSizes[VARIANT_MAPPING[variant]]};
`;

const Paragraph = styled(Text)`
  ${variantStyles};
`;

Paragraph.propTypes = {
  /**
   * Style variant. See storybook for examples.
   */
  variant: PropTypes.oneOf(['ui', 'longForm', 'tiny']).isRequired
};

Paragraph.defaultProps = {
  lineHeight: 'large'
};

export default Paragraph;
