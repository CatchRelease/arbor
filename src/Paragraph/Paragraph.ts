import { css, Theme } from '@emotion/react';
import styled from '@emotion/styled';
import type { LiteralUnion } from 'type-fest';

import Text from '../Text';
import { FontSize, fontSizeKeys } from '../theme/typography';

const VARIANT_MAPPING = {
  ui: 'size4',
  longForm: 'size5',
  tiny: 'size3'
} as const;

type Variant = 'ui' | 'longForm' | 'tiny';

const isFontSize = (value: string): value is FontSize =>
  fontSizeKeys.includes(value);

const variantStyles = ({
  theme,
  variant,
  fontSize
}: {
  theme: Theme;
  variant: Variant;
  fontSize?: LiteralUnion<FontSize, string>;
}) => {
  const fontSizeOverride =
    fontSize && isFontSize(fontSize) && theme.fontSizes[fontSize]
      ? theme.fontSizes[fontSize]
      : fontSize;

  return css`
    font-size: ${fontSizeOverride || theme.fontSizes[VARIANT_MAPPING[variant]]};
  `;
};

type Props = {
  fontSize?: LiteralUnion<FontSize, string>;
  variant: Variant;
};

const Paragraph = styled(Text)<Props>`
  ${variantStyles};
`;

Paragraph.defaultProps = {
  lineHeight: 'large'
};

export default Paragraph;
