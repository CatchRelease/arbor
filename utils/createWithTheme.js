import React from 'react';
import { CacheProvider } from '@emotion/core';
import createCache from '@emotion/cache';
import { create } from 'react-test-renderer';
import { ThemeProvider } from 'emotion-theming';
import theme from '../src/theme';

const myCache = createCache();
myCache.compat = true;

export default (Component) =>
  create(
    <CacheProvider value={myCache}>
      <ThemeProvider theme={theme}>{Component}</ThemeProvider>
    </CacheProvider>
  );
