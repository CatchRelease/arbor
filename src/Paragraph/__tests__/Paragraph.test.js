import React from 'react';
import { matchers } from 'jest-emotion';

import createWithTheme from '../../../tests/utils/createWithTheme';
import Paragraph from '../Paragraph';

expect.extend(matchers);

describe('<Paragraph />', () => {
  describe('props', () => {
    describe('variant', () => {
      ['ui', 'longForm', 'tiny'].forEach(variant => {
        it(`renders a ${variant} variant`, () => {
          const tree = createWithTheme(
            <Paragraph variant={variant}>Hello World</Paragraph>
          );

          expect(tree).toMatchSnapshot();
        });
      });
    });
  });
});
