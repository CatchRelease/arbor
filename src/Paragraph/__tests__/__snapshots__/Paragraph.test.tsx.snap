// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Paragraph /> fontSize properly renders a paragraph with custom font preset size 1`] = `
.emotion-0 {
  font-family: Motiva Sans;
  font-size: 1.25rem;
  line-height: 1.5;
  font-size: 1.25rem;
}

<p
  className="emotion-0"
  fontSize="size6"
>
  This is a paragraph
</p>
`;

exports[`<Paragraph /> fontSize properly renders a paragraph with custom pixel font size 1`] = `
.emotion-0 {
  font-family: Motiva Sans;
  font-size: 100;
  line-height: 1.5;
  font-size: 100;
}

<p
  className="emotion-0"
  fontSize="100"
>
  Size Button
</p>
`;

exports[`<Paragraph /> props variant renders a longForm variant 1`] = `
.emotion-0 {
  font-family: Motiva Sans;
  line-height: 1.5;
  font-size: 1rem;
}

<p
  className="emotion-0"
>
  Hello World
</p>
`;

exports[`<Paragraph /> props variant renders a tiny variant 1`] = `
.emotion-0 {
  font-family: Motiva Sans;
  line-height: 1.5;
  font-size: 0.8125rem;
}

<p
  className="emotion-0"
>
  Hello World
</p>
`;

exports[`<Paragraph /> props variant renders a ui variant 1`] = `
.emotion-0 {
  font-family: Motiva Sans;
  line-height: 1.5;
  font-size: 0.875rem;
}

<p
  className="emotion-0"
>
  Hello World
</p>
`;
